<a href="https://colab.research.google.com/github/Seventeee/CSC44800/blob/main/Csc_448_A4_Lun_Qu.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Name: Lun Qu <br>
Course: CSC 44800 Artificial Intelligence<br>
Assignment 4: Judging Flowers




<H2> Part 1 Loading the dataset </H2>

In [1]:
import numpy as np

import pandas as pd
from sklearn.datasets import load_iris

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

from sklearn.metrics import mean_squared_error, r2_score

from sklearn import svm
from sklearn.svm import SVC
from sklearn import datasets

from sklearn.neural_network import MLPClassifier



In [2]:
# Load the dataset (load remotely, not locally)
iris = load_iris()

# For read the data more easy, we might just put all data into pandas dataframe. 
# np.c_ is the numpy concatenate function
# which is used to concat iris['data'] and iris['target'] arrays 

# In the orignal data, we have two main array. 
# The one is called data set that record the the basic information of iris
# Sepal Length, Sepal Width, Petal Length and Petal Width.
# The second is called target which indicates the specials of iris. There are three 
# 0,1 and 2, they sepeartelly represent Iris-Setosa ; Iris-Versicolour; Iris-Virginica

iris_data_frame = pd.DataFrame(data= np.c_[iris['data'], iris['target']],
                     columns= iris['feature_names'] + ['target'])

In [3]:
# Output the first 15 rows of the data:
iris_data_frame.head(15)

sepal length (cm)  sepal width (cm)  ...  petal width (cm)  target
0                 5.1               3.5  ...               0.2     0.0
1                 4.9               3.0  ...               0.2     0.0
2                 4.7               3.2  ...               0.2     0.0
3                 4.6               3.1  ...               0.2     0.0
4                 5.0               3.6  ...               0.2     0.0
5                 5.4               3.9  ...               0.4     0.0
6                 4.6               3.4  ...               0.3     0.0
7                 5.0               3.4  ...               0.2     0.0
8                 4.4               2.9  ...               0.2     0.0
9                 4.9               3.1  ...               0.1     0.0
10                5.4               3.7  ...               0.2     0.0
11                4.8               3.4  ...               0.2     0.0
12                4.8               3.0  ...               0.1     0.0
13                4.3               3.0  ...               0.1     0.0
14                5.8               4.0  ...               0.2     0.0

[15 rows x 5 columns]

In [4]:
iris_data_frame.dtypes

sepal length (cm)    float64
sepal width (cm)     float64
petal length (cm)    float64
petal width (cm)     float64
target               float64
dtype: object

In [5]:
iris_data_frame.describe()

sepal length (cm)  sepal width (cm)  ...  petal width (cm)      target
count         150.000000        150.000000  ...        150.000000  150.000000
mean            5.843333          3.057333  ...          1.199333    1.000000
std             0.828066          0.435866  ...          0.762238    0.819232
min             4.300000          2.000000  ...          0.100000    0.000000
25%             5.100000          2.800000  ...          0.300000    0.000000
50%             5.800000          3.000000  ...          1.300000    1.000000
75%             6.400000          3.300000  ...          1.800000    2.000000
max             7.900000          4.400000  ...          2.500000    2.000000

[8 rows x 5 columns]

<H2>About the dataset</H2>
As we expained previous, there are three species of iris. The experiment the scientist did is trying to Collected data to quantify the morphological variation of iris flowers in three related species. We will use those data to build our different classification models. Thus the feature X should be iris['data'] contains sepal length (cm), sepal width (cm), petal length (cm) and petal width (cm). The label y will be the target, the number indicates iris class.

<H2>Part 2: Split the dataset into train and test</H2>

In [6]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = np.array(iris_data_frame.drop(['target'], 1))
y = np.array(iris_data_frame['target'])

In [7]:
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# numpy.random. RandomState() This is a pseudorandom number generator.
# we can also make random_state be a exactly number, so we can control the model we get.
rng = np.random.RandomState(0)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=rng)

<H2>Part 3: Logistic Regression</H2>

In [8]:
# i. Use sklearn to train a LogisticRegression model on the training set
LR_model = LogisticRegression()
#Training the model
LR_model.fit(X_train, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [9]:
# ii. For a sample datapoint, predict the probabilities for each possible class
# We can test our model by the first point in test set
# the first predict_proba() return estimates for all classes are ordered by the label of classes.
# since we have three labal, our return will be a array have three numbers
probabilities = LR_model.predict_proba(X_test[:1, :])
print('For the iris sample:', X_test[:1, :])
print('class 0(Setosa):',probabilities[0][0],'|class 1(Versicolour):',probabilities[0][1],'|class 2(Virginica):',probabilities[0][2])


For the iris sample: [[5.8 2.8 5.1 2.4]]
class 0(Setosa): 8.233434079212403e-05 |class 1(Versicolour): 0.05049456070178681 |class 2(Virginica): 0.949423104957421


In [10]:
# iii. Report on the score for Logistic regression model
LR_confidence = LR_model.score(X_test, y_test)
print('The score of our Logistic Regression Model is',LR_confidence)


The score of our Logistic Regression Model is 1.0


By the function, we get the score of our Logistic Regression Model is perfectly 1, which mean our model always correctly predict the right class of iris 

In [12]:
# iv. Extract the coefficents and intercepts for the boundary line(s)
# The coefficients and intercepts:
print('Coefficients: \n', LR_model.coef_)
print('Intercepts: \n', LR_model.intercept_)

Coefficients: 
 [[-0.41087776  0.92630693 -2.43464513 -1.04770079]
 [ 0.40315439 -0.32622985 -0.11363366 -0.88710555]
 [ 0.00772337 -0.60007709  2.5482788   1.93480634]]
Intercepts: 
 [  9.46338562   2.44534766 -11.90873328]


Obviously, we will have three boundary line for each class. At above, the first list the coefficients list is the coefficients of the fisrt boundary line for class 0. Each conefficient relatives to one features, we have four features which perfectly match the coefficients. And the intercepts of the first boundary line is 9.4634. The other two boundary line follow the same format.

<H2>Part 4: Support Vector Machine</H2>

In [13]:
# i. Use sklearn to train a Support Vector Classifier on the training set
SVC_model = SVC(kernel = 'linear', random_state = rng, probability = True)
SVC_model.fit(X_train, y_train)

SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='scale', kernel='linear',
    max_iter=-1, probability=True,
    random_state=RandomState(MT19937) at 0x7FD6EABCB490, shrinking=True,
    tol=0.001, verbose=False)

In [14]:
# ii. For a sample datapoint, predict the probabilities for each possible class
probabilities = SVC_model.predict_proba(X_test[:1, :])
print('For the iris sample:', X_test[:1, :])
print('class 0(Setosa):',probabilities[0][0],'|class 1(Versicolour):',probabilities[0][1],'|class 2(Virginica):',probabilities[0][2])


For the iris sample: [[5.8 2.8 5.1 2.4]]
class 0(Setosa): 0.009725221987402229 |class 1(Versicolour): 0.006937259777277787 |class 2(Virginica): 0.98333751823532


In [15]:
# iii. Report on the score for the SVM, what does the score measure?
SVC_confidence = SVC_model.score(X_test, y_test)
print('The score of our Support Vector Classifier is',SVC_confidence)


The score of our Support Vector Classifier is 1.0


Once again, we got the score of our support vector machine is also 1. If we compare the second part, the probalities of each possible class of the same is very similarly to the last we got from Logistic regression model. The class 3 has the highest ratio. 

<H2>Part 5: Neural Network</H2>

In [16]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
# ‘lbfgs’ is an optimizer in the family of quasi-Newton methods.
NN_model = MLPClassifier(solver='lbfgs',tol = 1e-6,hidden_layer_sizes=(5,4),activation="logistic",alpha=1e-1,max_iter=1000,random_state=rng)
NN_model.fit(X_train, y_train)


MLPClassifier(activation='logistic', alpha=0.1, batch_size='auto', beta_1=0.9,
              beta_2=0.999, early_stopping=False, epsilon=1e-08,
              hidden_layer_sizes=(5, 4), learning_rate='constant',
              learning_rate_init=0.001, max_fun=15000, max_iter=1000,
              momentum=0.9, n_iter_no_change=10, nesterovs_momentum=True,
              power_t=0.5, random_state=RandomState(MT19937) at 0x7FD6EABCB490,
              shuffle=True, solver='lbfgs', tol=1e-06, validation_fraction=0.1,
              verbose=False, warm_start=False)

In [17]:
# ii. For a sample datapoint, predict the probabilities for each possible class
probabilities = NN_model.predict_proba(X_test[:1, :])
print('For the iris sample:', X_test[:1, :])
print('class 0(Setosa):',probabilities[0][0],'|class 1(Versicolour):',probabilities[0][1],'|class 2(Virginica):',probabilities[0][2])



For the iris sample: [[5.8 2.8 5.1 2.4]]
class 0(Setosa): 0.0017239002747314013 |class 1(Versicolour): 0.006254101953466275 |class 2(Virginica): 0.9920219977718023


In [18]:
# iii. Report on the score for the Neural Network, what does the score measure?
NN_confidence = NN_model.score(X_test, y_test)
print('The score of our Neutral Network Model is',NN_confidence)

The score of our Neutral Network Model is 1.0


In my person neural network model, I set tol = 0.000001, which is the threshold of error, our model will stop training when it is less than this value. hidden_layer_sizes: set two hidden layer inside the network, the first hidden layer has 5 nodes, and the second has 4 nodes. The activition indicates the activation function of the hidden layer to be logistic. Alpha is 0.1, becuase I don't want it too small, that we need too many iteration to find optimal solution.

<H2>Part 6: K-Nearest Neighbors</H2>

In [19]:
from sklearn.neighbors import KNeighborsClassifier

# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier
knn = KNeighborsClassifier(n_neighbors=10)
knn.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class

probabilities = knn.predict_proba(X_test[10:11, :])
print('For the iris sample:', X_test[10:11, :])
print('class 0(Setosa):',probabilities[0][0],'|class 1(Versicolour):',probabilities[0][1],'|class 2(Virginica):',probabilities[0][2])

# iii. Report on the score for kNN, what does the score measure?
confidence = knn.score(X_test, y_test)
print('The score of our K-Nearest Model is',confidence)


For the iris sample: [[6.1 2.6 5.6 1.4]]
class 0(Setosa): 0.0 |class 1(Versicolour): 0.2 |class 2(Virginica): 0.8
The score of our K-Nearest Model is 1.0


The K-Nearest Neighbors model will define the class of the point by nearest n point. It is also a good idea to predict the new point give to us, because the point share the same class should have the similar features. 

<H2>Part 7: Conclusions and takeaways</H2>

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?<Br>
    Base the previous work, the all scores of today's meachine model is 1, which means all models are able to define the point correctly. For, i think the logistic regression and the Neutal network model make more sense to me. The logistic regression constructs linear boundaries to classfer multiple classes. The boundary line gives me a good visualization. The building of neutal network model, we gradually get closer to optimal answer. 
    K-NN algorithm is also very simple to understand and equally easy to implement. But I feel it too dependent on the balance of input data. If one class only get few input data, then the model will ultimately give a lot of preference to majority set. This might result in getting the less common class wrongly classified.  
    